text/0034-xcm-absolute-location-account-derivation.md

```diff
@@ -0,0 +1,199 @@
+# RFC-34: XCM Absolute Location Account Derivation
+
+|                 |                                                                                             |
+| --------------- | ------------------------------------------------------------------------------------------- |
+| **Start Date**  | 05 October 2023                                                                             |
+| **Description** | XCM Absolute Location Account Derivation                                                    |
+| **Authors**     | Gabriel Facco de Arruda                                                                     |
+
+## Summary
+
+This RFC proposes a change to the `WithComputedOrigin` XCM barrier and a backwards-compatible change to the `DescribeFamily` MultiLocation descriptor with the end goal of enabling the use of Universal Locations for XCM MultiLocation to AccountId conversion, which allows the use of absolute locations to maintain the same derivation result in any runtime, regardless of it's position in the family hierarchy.
+
+## Motivation
+
+These changes would allow protocol builders to leverage absolute locations to maintain the exact same derived account address across alls network in the ecosystem, thus enhancing user experience.
+
+One such protocol, that is the original motivation for this proposal, is InvArch's Saturn Multisig, which gives users a unifying multisig and DAO experience across all XCM connected chains.
+
+## Stakeholders
+
+- Ecosystem developers
+
+## Explanation
+
+This proposal requires the modification of two XCM types defined in the `xcm-builder` crate: The `WithComputedOrigin` barrier and the `DescribeFamily` MultiLocation descriptor.
+
+This proposal will go through the actual code changes that should be made, however the code provided will only serve to illustrate the goal of the changes and the actual implementation code is subject to further discussions.
+
+#### WithComputedOrigin
+
+The `WtihComputedOrigin` barrier serves as a wrapper around other barriers, consuming origin modification instructions and applying them to the message origin before passing to the inner barriers. One of the origin modifying instructions is `UniversalOrigin`, which serves the purpose of signaling that the origin should be a Universal Origin that represents the location as an absolute path with the interior prefixed by the `GlobalConsensus` junction.
+
+The change that needs to be made here is to remove the current relative location conversion that is made and replace it with an absolute location, represented as `{ parents: 2, interior: [GlobalConsensus(...), ...] }`.
```

Comment:
An absolute location contains no parent junctions. In fact, the code change you outlined here does pretty much the exact same thing as what it did before, when it was using `relative_to`.

Reply:
What I understood was that an absolute location was represented as parents 2 and an interior prefixed with GlobalConsensus, but since I can't find documentation of this anywhere I just had to use my assumptions. Should absolute locations have parents 0 + GlobalConsensus interior prefix then?

From what I tested, the relative_to code would output a relative MultiLocation that's relative to the local location, for example: parents 0/interior here on the relay and parents 1/interior here in paras, which is not at all what I'm trying to achieve, and even if I'm somehow getting this wrong, the method is quite literally called relative_to, if it's meant to return an absolute location then it's very incorrectly named.

Reply:
> What I understood was that an absolute location was represented as parents 2 and an interior prefixed with GlobalConsensus

This is incorrect. You are describing a relative location from a foreign parachain to reach your network.

> even if I'm somehow getting this wrong, the method is quite literally called relative_to, if it's meant to return an absolute location then it's very incorrectly named.

It is not intended to return an absolute location at all.

The biggest difference between an absolute location and a relative location is that the former has an implicit ultimate root location. By definition, it does not contain any parents. Such a concept did not exist before in XCM v2, until in v3 where we added a new pseudo-location called the `UniversalConsensus`. It is pseudo because it does not in fact exist, but is merely a means to describe the parent that all global consensus systems share. There's a couple of implications from this:

1. There should never be a location that points directly to the `Universal

Reply:
I still fail to understand how absolute locations are supposed to be used at all, as traits like `ShouldExecute` and `DescribeLocation` all take `MultiLocation`, so I don't understand why moving the absolute location through these traits as a parents: 2 MultiLocation with interior prefixed by GlobalConsensus is such a big deal, considering this wrapper `MultiLocation` is just that, a wrapper to separate relative locations from absolute ones, as when it comes time to use this absolute location, only the interior is actually used.

The flow originally proposed in this RFC and that still makes the most sense to me is:
1. Chain receives XCM with prefix instruction of `UniversalLocation`
2. `WithComputedOrigin` barrier consumes the prefix instruction and morphs the location to `parents: 2, interior: [GlobalConsensus, ...]`
3. `DescribeFamily` then matches locations with the pattern `parents: 2, interior: [GlobalConsensus, ...]` discards everything but the interior and describes that interior as an absolute location

Since everything from the barrier's `ShouldExecute` trait to the AccountId converter's `ConvertLocation` trait works with `MultiLocation` inputs, I don't see why we can't move this absolute location internally between these traits as a specific `MultiLocation` pattern that describes absolute locations.

Reply:
Would this be a more correct approach for implementing absolute locations?

In `WithComputedOrigin`
```rust
UniversalOrigin(new_global) => {
    actual_origin =
        X1(*new_global)
            .within_global(actual_origin)
            .map_err(|_| ProcessMessageError::Unsupported)?
            .into();
}
```

And in `DescribeFamily`
```rust
match (l.parents, l.interior.first()) {
    (0, Some(GlobalConsensus(network_id))) => {
```